### ISSUE TITLE
[feature]: Disable redirects

### ISSUE THREAD

**Issue by KaKi87** [hoppscotch/hoppscotch#2095]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Disable redirects

### Why should this be worked on?

Disable redirects

**Comment by iamajar** [comment]
I need the feature too.
Just like the setting of Postman - "Automatically follow redirects : off".

**Comment by DarkCWK** [comment]
I need the feature too.
Sometimes we just need to see `302 Found`

**Comment by AndrewBastin** [comment]
> Sometimes we just need to see 302 Found

That is exactly the problem, since we are backed by the browser APIs at the end of the day, due to the Atomic HTTP Redirect Handling policy in the `fetch` implementation, we will not be able to get access to the redirect information. If we ask the browser to disable redirect, it will just send us out a generic network error where we can't guarantee whether it was a connectivity issue, invalid content or even a redirect.

See spec: https://fetch.spec.whatwg.org/#atomic-http-redirect-handling

This may be possible to be implemented in Proxyscotch tho.

I am keeping this issue open so we can eventually implement it when we are building the native app.

**Comment by KaKi87** [comment]
> This may be possible to be implemented in Proxyscotch tho.

Exactly, please.
> we are building the native app

By native, do you mean non-Electron ?

**Comment by AndrewBastin** [comment]
> Exactly, please.

Please create an issue in github.com/hoppscotch/proxyscotch so that can be tracked.

> By native, do you mean non-Electron ?
Absolutely not, we are planning to go with Tauri, but this is still long ways off, I don't expect it to even be in an Alpha stage till 2023.

**Comment by KaKi87** [comment]
> Please create an issue in github.com/hoppscotch/proxyscotch so that can be tracked.

Oh I'm sorry, I thought that was the extension.

Could the extension handle this ?
> we are planning to go with Tauri

Nice !
> this is still long ways off, I don't expect it to even be in an Alpha stage till 2023

What would take that long ?
Tauri's HTTP methods are easy enough to use so that I made an [`axios` adapter](https://git.kaki87.net/KaKi87/axios-tauri-adapter) for it.
Would you like a `fetch` adapter ?

**Comment by AndrewBastin** [comment]
> Could the extension handle this ?
AFAIK, I don't think so. Extension still uses the browser APIs for the most part.

> What would take that long ?
Tauri's HTTP methods are easy enough to use so that I made an [axios adapter](https://git.kaki87.net/KaKi87/axios-tauri-adapter) for it.
Would you like a fetch adapter ?


Thank you for your suggestion (your code looks awesome! :heart:).

We are still a small team and we want to keep the web app as the main priority still. We believe the web app is the future for the most part and native is just a way to keep the app not limited due to browser issues meanwhile. We want to spend a lot of time, just figuring out how to make the development part really easy in a way that we can just build the web app and the native app together at the same time in a way that it is really clean because we think it is a worthwhile investment. So, I am working towards that for now.

You can DM me if you are interested in continuing the conversation as it is getting a bit off-topic. You can find me on the Hoppscotch Discord/Telegram (link in README).

**Comment by drilus** [comment]
I need this feature as well.

**Comment by Daanoz** [comment]
As an alternative, would it be possible to expose the url where the request redirected to, in the `pw.response` object so it could be used in the scripts? https://docs.hoppscotch.io/documentation/features/scripts#pwresponse

**Comment by KaKi87** [comment]
That's not an alternative because it wouldn't provide access to everything else the redirecting server returns (code, headers, even body), nor handle multiple redirections.

**Comment by Daanoz** [comment]
True, it's not a full solution, but it might be something that can be more easily realized on a shorter timeframe.

In my usecase, I have an oAuth service with PKCE flow which I'm trying to validate. The oAuth spec defines one of the requests has to be a redirect, which includes a code token as parameter in the url. With the current client, there is simply no way to access that code value, not via the UI and not via scripting.

**Comment by adilek** [comment]
This feature is important for me and my team as well.

**Comment by bttger** [comment]
Why not simply add the `redirect: "error"` prop if users ask to disable redirects? I don't care about the location and status code. If really necessary I could just use another tool (e.g. httpie where redirects are always disabled). The fetch request would then just return `TypeError: Failed to fetch`. And either the user decides whether this has happened due to a redirect (since he/she disabled them intentionally) or hoppscotch does another request to test if the failure was due to the redirect. The latter would probably be preferred.


Here's how i tested it based on the [MDN fetch redirect docs](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirected#disallowing_redirects):
```html
<!DOCTYPE html>
<html>
  <head>
    <title>Test Disabling Redirects</title>
  </head>
  <body>
    <h1>Test Disabling Redirects</h1>

    <script>
      window.addEventListener("load", function () {
        fetch(
          "http://raw.githubusercontent.com/bttger/wake-up-light/my-config/sunrise.json",
          { redirect: "error" }
        )
          .then((response) => {
            console.log(response);
            for (var pair of response.headers.entries()) {
              console.log(pair[0] + ": " + pair[1]);
            }
            return response.json();
          })
          .then((json) => console.log(json))
          .catch((error) => console.log(error));
      });
    </script>
  </body>
</html>

```

**Comment by KaKi87** [comment]
> I don't care about the location and status code.

I do.

**Comment by bttger** [comment]
> I do.

Then use another tool - it's not possible to get that info in a browser. But I showed a way how to disable redirects and to at least tell whether a redirect was responsible for the failure (if a request failed). I believe this would already help a lot.

**Comment by KaKi87** [comment]
> it's not possible to get that info in a browser

That's what the proxy and the desktop app are for.

**Comment by bttger** [comment]
> That's what the proxy and the desktop app are for.

The proxy is a browser extension with the same limitation and the desktop app doesn't provide that feature either. What point do you want to make?

**Comment by KaKi87** [comment]
> The proxy is a browser extension with the same limitation

No, [the proxy is the proxy](https://github.com/hoppscotch/proxyscotch), and [the browser extension is the browser extension](https://github.com/hoppscotch/hoppscotch-extension).

Therefore, since not running in browser, the proxy could disable redirects.

> the desktop app doesn't provide that feature either

Yes [it does provide that feature](https://tauri.app/v1/api/js/http/), so it could disable redirects as well.

**Comment by bttger** [comment]
Sorry, I didn't know there was an additional proxy. Had a quick look into it but personally I have no use case to use the proxy so I'd like to avoid the hassle (especially when working with WSL/docker and thinking that the USP of Hoppsotch is that I can just open the tool quickly on my work laptop without installing anything).

And yes, of course does Tauri support disabling of redirects but I was talking about the desktop app - it doesn't provide the disabling of redirects. And I still believe that my proposal would be a nice improvement over the current situation that would be easy to implement. Yes, it wouldn't show the status code but it would tell you that the request has failed due to a redirect.

**Comment by yannleretaille** [comment]
Since the desktop app is using tauri now, shouldn't it be possible to add an option to requests that sets [curl::Easy::max_redirections](https://docs.rs/curl/latest/curl/easy/struct.Easy.html#method.max_redirections) to `0` to the [native interceptor/relay](https://github.com/hoppscotch/hoppscotch/blob/main/packages/hoppscotch-relay/src/relay.rs)? 

Just hit the same issue testing an auth endpoint earlier, where I just needed to see the redirect status code, location headers etc. to confirm that it's working as intended and instead got a "Could not send request" error.

**Comment by macmessa** [comment]
Is this in the backlog?

**Comment by carlos-dubon** [comment]
i have the opposite problem. i want to enable them, but i cant find the toggle

**Comment by KaKi87** [comment]
> i want to enable them

The reason why this issue is a feature request for disabling redirects is that redirects are enabled by default and cannot be disabled.

**Comment by carlos-dubon** [comment]
> > i want to enable them
> 
> The reason why this issue is a feature request for disabling redirects is that redirects are enabled by default and cannot be disabled.

Yeah I'm not trying to be odd or something. It's just that mine isn't redirecting. Maybe I'll open another issue

**Comment by KaKi87** [comment]
Okay, though that's likely an issue on your API, I'm afraid. I'd suggest getting raw output using cURL, feeding it to an AI, along with the redirect you expected, and asking it why it doesn't.

**Comment by jamesgeorge007** [comment]
Hi, the ability to disable redirects has been added to the Desktop App and [Agent](https://docs.hoppscotch.io/documentation/features/interceptor#hoppscotch-agent) micro application for use in the web app as part of the [latest release](https://github.com/hoppscotch/hoppscotch/releases/tag/2025.10.0). Please feel free to share any feedback.

<img width="1494" height="435" alt="Image" src="https://github.com/user-attachments/assets/00cddcdd-976c-4614-9eed-9c88a122dba2" />

<img width="1476" height="680" alt="Image" src="https://github.com/user-attachments/assets/2b89383f-f38a-4bf1-b077-33aee0adc125" />
